contracts/exchange/contracts/test/ReentrantERC20Token.sol

```diff
@@ -52,137 +50,280 @@ contract ReentrantERC20Token is
+        for (uint8 i = 0; i != count; i++) {
+            amounts[i] = orders[i].takerAssetAmount;
+        }
+    }
+
+    function sumTakerFillAmounts(
+        LibOrder.Order[] memory orders
+    )
+        internal
+        pure
+        returns (uint256 total)
+    {
+        uint256 count = orders.length;
+        total = 0;
+        for (uint8 i = 0; i != count; i++) {
+            total += orders[i].takerAssetAmount;
+        }
+    }
+
+    /// @dev Generate a random address.
+    function getRandomAddress()
+        internal
+        view
+        returns (address)
+    {
+        return address(
+            bytes20(
+                keccak256(
+                    abi.encodePacked(
+                        blockhash(block.number - 1), now)
+                    )
+                )
+            );
+    }
+
+    /// @dev Create empty wallet-verified signatures.
+    function createWalletSignatures(
+        uint8 count
+    )
+        internal
+        returns (bytes[] memory signatures)
+    {
+        signatures = new bytes[](count);
+        for (uint i = 0; i != count; i++) {
+            signatures[i] = new bytes(66);
+            signatures[i][65] = bytes1(uint8(0x4));
+        }
+    }
+
+    /// @dev Create asset data that points to this ERC20 contract.
+    function createAssetData()
+        internal
+        view
+        returns (bytes memory assetData)
+    {
+        // ERC20 AssetData is encoded as follows:
+        //
+        // | Area     | Offset | Length  | Contents                            |
+        // |----------|--------|---------|-------------------------------------|
+        // | Header   | 0      | 4       | function selector                   |
+        // | Params   |        | 1 * 32  | function parameters:                |
+        // |          | 4      | 12 + 20 |   1. token address                  |
+
+        // We'll build a version of the AssetData in memory in this layout
+        //
+        // | Area     | Offset | Length  | Contents                            |
+        // |----------|--------|---------|-------------------------------------|
+        // | Size     | 0      | 32      | length of the bytes array           |
+        // | Header   | 32     | 4       | function selector                   |
+        // | Params   |        | 1 * 32  | function parameters:                |
+        // |          | 36     | 12 + 20 |   1. token address                  |
+
+        assetData = new bytes(36);
+        assembly {
+            // First 32 bytes of a bytes array holds the size (36), which should
+            // already be set when creating the bytes array.
+
+            // First we encode the the function selector for the
+            // ERC20 asset proxy: 0xf47261b.
+            mstore(add(assetData, 32), 0xf47261b000000000000000000000000000000000000000000000000000000000)
+            // Then we encode parameter 1, the token address.
+            mstore(add(assetData, 36), address)
```

Comment:
Just a note -- when storing a payload ≤ 32 bytes we typically apply a mask to zero out the unused portion of the word. What we have here is probably okay since (i) `assetData` is allocated with `new` directly above, and (ii) historically the compiler has been very liberal with masks so it probably applies one to `address`. That said, neither (i) nor (ii) completely guarantee the word will be zeroed out.

Reply:
I can make this change, np.